File: Logres/AStar/src/framework/AStar.java
package framework;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;

public class AStar {

	private ArrayList<AsNode> OPEN;
	private HashMap<Integer, AsNode> states;
	private AsStateGenerator generator;
	public AsNode goalNode;
	public int expandedNodes = 0;

	public AStar(AsStateGenerator generator, AsState initialState, AsState goalState) {

		//	Initialize helper classes.
		this.generator = generator;
		OPEN = new ArrayList<AsNode>();
		states = new HashMap<Integer, AsNode>();
		
		//	Ensure goalstate is properly generated.
		goalState.setHash(generator.generateHash(goalState));
		goalState.setH(this.generator.calculateH(goalState, goalState));
		this.goalNode = new AsNode(goalState);
		
		//	Handle and prepare root node
		initialState.setHash(generator.generateHash(initialState));
		initialState.setG(0);
		initialState.setH(generator.calculateH(initialState, goalState));
		AsNode root = new AsNode(initialState);
		states.put(root.getHash(), root);
		OPEN.add(root);
	}
	
	/***
	 * This function runs the A* algorithm, for the entered parameters.
	 * @return boolean, sucess state of the algorithm.
	 */
	public boolean run() {
		boolean success = false;
		AsNode currentNode;
		//This is the agenda loop
		do {
			currentNode = OPEN.remove(0);
			if (currentNode.equals(goalNode)){
				goalNode = currentNode;
				success = true; 
				break;
			}
			
			ArrayList<AsState<?>> children = generator.getChildren(currentNode.getState());
			currentNode.OPEN = false;
			expandedNodes++;
			for (AsState<?> child : children) {
				
				//Check if child state already exists, and set/treat it accordingly.
				AsNode childNode;
				if(states.containsKey(child.getHash())) {
					childNode = states.get(child.getHash());
				}
				else {
					childNode = new AsNode(child);
				}
				
				//Calculate distance from parent to generated child state, and add it to it's children
				currentNode.addChild(childNode, child.getG() - currentNode.getState().getG());
				
				//If a new node was found, evaluate and add it to OPEN using binary insertion. This ensures the OPEN is sorted in O(log(n)) time. 
				if(!states.containsKey(child.getHash())) {
					attachAndEval(currentNode, childNode);
					states.put(childNode.getHash(), childNode);
					OPEN.add(binarySearch(OPEN, 0, OPEN.size(), childNode.getF()), childNode);
				}
				//If node was already made, check to see if it has a new optimal path to it, and propagate if it's children were already handled.
				else if (childNode.getState().getG() > currentNode.getState().getG() + currentNode.getArcCost(childNode)) {
					System.out.println("Updated node");
					attachAndEval(currentNode, childNode);
					if (!childNode.OPEN) {
						System.out.println("Propagatin");
						propagate(childNode);
					}
					else if(childNode.OPEN) {
						System.out.println("Reputting child");
						OPEN.remove(childNode);
						OPEN.add(binarySearch(OPEN, 0, OPEN.size(), childNode.getF()), childNode);
					}
				}
			}
		} while(OPEN.size() > 0);
		System.out.println(success);
		return success;
	}
	
	private void attachAndEval(AsNode parent, AsNode child) {
		child.setParent(parent);
		child.getState().setH(generator.calculateH(child.getState(), goalNode.getState()));
	}
	
	private void propagate(AsNode parent) {
		double G = parent.state.getG();
		for (AsNode child : parent.getChildren()) {
			double toChild = parent.childCost.get(child);
			double childG = child.state.getG();
			
			if (childG > G + toChild) {
				child.setParent(parent);
				child.state.setG(G + toChild);
				if(child.OPEN) {
					OPEN.remove(child);
					OPEN.add(binarySearch(OPEN, 0, OPEN.size(), child.getF()), child);
				}
				propagate(child);
			}
		}
	}
	
	/***
	 * You must run the run() function before calling this this function, else it returns and empty goal state.
	 * @return Goal node.
	 */
	public AsNode getSolution() {
		return goalNode;
	}
	
	
	//This function is used for binary search in an ArrayList.
	private static int binarySearch(ArrayList<AsNode> a, int start, int end, double val){
		   
	    if(start == end){
	    	return start;
	    }
	    int middle = (start + end)/2;
	    if(val <= a.get(middle).getF()){
	        return binarySearch(a, start, middle, val);
	    }
	    else{
	        return binarySearch(a, middle + 1, end, val); 
	    }
	}
}

File: AIP/PStar/AStar.py
from ASNode import ASNode


MODE_BEST_FIRST = 0
MODE_DEPTH_FIRST = 1
MODE_BREADTH_FIRST = 2

class AStar(object):
	
	def __init__(self, renderer = None):
		self.OPEN = None
		self.nodes = None
		self.start = None
		self.goal = None
		self.success = False
		self.renderer = renderer
		self.mode = 0

	def initialize(self, start, goal):
		self.success = False

		self.start = start
		self.start.hash = start.generateHash()
		self.start.calculateHeuristic(goal)
		
		self.goal = goal
		self.goal.hash = goal.generateHash()
		
		self.OPEN = []
		self.nodes = {}
		
		self.OPEN.append(start)
		self.nodes[start.hash] =  start
		
	def setMode(self, mode):
		self.mode = mode
	
	def insertIntoOpen(self, node):
		if(self.mode == 0):
			#BEST-FIRST
			self.OPEN.insert(self.binarySearch(self.OPEN, 0, len(self.OPEN), node), node)
		elif(self.mode == 1):
			#DEPTH-FIRST
			self.OPEN.insert(0, node)
		else:
		#BREADTH-FIRST
			self.OPEN.append(node)
		
		
	def binarySearch(self, list, start, end, node):
		if (start == end):
			return start
		
		middle = int((start + end) / 2)
		if (node.getF() <= list[middle].getF()):
			return self.binarySearch(list, start, middle, node)
		else:
			return self.binarySearch(list, middle + 1, end, node)
		
	
	def attachAndEval(self, parent, child):
		child.h = child.calculateHeuristic(self.goal)
		child.g = parent.g + parent.childCost.get(child.hash)
		child.parent = parent
		
	def propagate(self, node):
		g = 0.0
		g = node.g
		
		for child in node.children:
			newG = g + node.childCost[child.hash]
			print(newG - child.g)
			if (newG < child.g):
				print("Found Shorter path")
				child.parent = node
				child.g = newG
				
				if(not child.expanded):
					self.OPEN.remove(child)
					self.insertIntoOpen(child)
					
				propagate(child)
			

	def render(self, node):
		if (not self.renderer is None):
			self.renderer.render(self, node)
			return
			
	def run(self):
		expandedNodes = 0
		print("running")
		if(self.OPEN is None or self.nodes is None or self.start is None or self.goal is None or self.success is True):
			print("Algorithm not properly initialized")
			return
		
		node = None
		
		while(len(self.OPEN) > 0):
			node = self.OPEN.pop(0)
			print("Popped node: {}\tHash: {}\tG: {}\tH: {}\tF: {}".format(expandedNodes, node.hash, node.g, node.h, node.getF()))
			#Check if node was goalnode
			if(node.hash == self.goal.hash):
				print("Success!")
				self.success = True
				self.goal = node
				return node
			
			node.expanded = True
			
			#Expand
			children = node.generateChildren()
			
			for i in range(len(children)):
				freshNode = False
				child = children[i]
				
				#Check if childNode exists
				childNode = self.nodes.get(child.hash)
				
				if (childNode is None):
					#node not found yet, make new one.
					childNode = child
					freshNode = True
				
				node.addChild(childNode, child.g - node.g)
				
				if(freshNode):
					self.attachAndEval(node, childNode)
					self.nodes[childNode.hash] = childNode
					self.insertIntoOpen(childNode)

				elif(childNode.g > node.g + node.childCost.get(childNode.hash)):
					self.attachAndEval(node, childNode)
					
					if(childNode.expanded):
						self.propagate(childNode)
					else:
						self.OPEN.remove(childNode)
						self.insertIntoOpen(childNode)	

			child.expanded = True
			expandedNodes += 1		
			self.render(node)
	